chalk/types.py

```diff
@@ -1,35 +1,49 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, List, Optional, Tuple
-
-from svgwrite import Drawing
-from svgwrite.base import BaseElement
+from typing import TYPE_CHECKING, Any, List, Optional, Protocol, Tuple
 
 import chalk.transform as tx
 from chalk.envelope import Envelope
 from chalk.style import StylableProtocol, Style
 from chalk.trace import Trace
-from chalk.transform import V2
+from chalk.transform import P2, V2, Transformable
 
 if TYPE_CHECKING:
-    from chalk.shape import Shape
-    from chalk.visitor import A, DiagramVisitor
-
-__all__ = ["BaseElement", "Drawing"]
-PyLatexElement = Any
-PyLatex = Any
-PyCairoContext = Any
-PyCairoSurface = Any
+    from chalk.visitor import A, DiagramVisitor, ShapeVisitor
+
 Ident = tx.Affine.identity()
 
 
-class Diagram(StylableProtocol, tx.TransformableProtocol):
-    def get_envelope(self, t: tx.Affine = Ident) -> Envelope:
+class Enveloped(Protocol):
+    def get_envelope(self) -> Envelope:
+        ...
+
+
+class Traceable(Protocol):
+    def get_trace(self) -> Trace:
         ...
 
-    def get_trace(self, t: tx.Affine = Ident) -> Trace:
+
+class Shape(Enveloped, Traceable, Protocol):
+    def accept(self, visitor: ShapeVisitor[A], **kwargs: Any) -> A:
         ...
 
+
+class SegmentLike(Enveloped, Traceable, Transformable):
+    @property
+    def p(self) -> P2:
+        ...
+
+    @property
+    def q(self) -> P2:
+        ...
+
+    dangle: float
```

Comment:
This is minor, but I wonder whether `dangle` should be a common attribute to `Segment` and `ArcSegment`, given that it's not really relevant for the `Segment` subclass. I assume that you have included it for its use in the `arrow.py` module, but maybe it can be removed if we explicitly check that the segment is an `ArcSegment` prior to rotating (although I understand this is not a particularly elegant solution):
```diff
diff --git a/chalk/arrow.py b/chalk/arrow.py
index ecd0422..7e56984 100644
--- a/chalk/arrow.py
+++ b/chalk/arrow.py
@@ -94,8 +94,8 @@ def arrow(length: float, style: ArrowOpts = ArrowOpts()) -> Diagram:
             P2(0, 0), P2(l_adj, 0), style.arc_height
         )
         shaft = Path([segment]).stroke()
-        φ = segment.dangle
-        arrow = arrow.rotate_rad(φ)
+        if isinstance(segment, ArcSegment):
+            arrow = arrow.rotate_rad(segment.dangle)
```

Reply:
yeah, that's a better solution.

Reply:
Actually is there a nice way to do Union types in python now? Maybe segment should just be `Segment | ArcSegment` since I don't think there are going to be more segment types.

Reply:
Yes, I was also thinking of using a [Union type](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#union-types), but I was not sure if would break anything.

Reply:
No that sounds right to me. Perhaps `Shapes` should also be a `Union` since they need to be closed class now for rendering. Visitor is perhaps too fancy.

Reply:
Yes, you are probably right; I guess an upside of the current (visitor-based) implementation is that the `Shape` rendering implementation is a bit more consistent with the `Diagram` one :slightly_smiling_face: I'm fine with either option.

Reply:
Okay, moved Segment to be a union. Left shape as is.